### PR TITLE
openlineage, sftp: add OpenLineage support for SFTPOperator

### PR DESCRIPTION
**Pull request by mobuchowski**
This PR adds OpenLineage metadata support for SFTPOperator.

### CODE REVIEW

**Comment by pankajkoti** [comment]
cc: @utkarsharma2 could you please review this?

File: airflow/providers/sftp/operators/sftp.py [hunk 1]
```diff
@@ -188,3 +191,85 @@ def execute(self, context: Any) -> str | list[str] | None:
...
+
+        local_datasets = [
+            Dataset(namespace=self._get_namespace(scheme, local_host, None, path), name=path)
+            for path in local_filepath
+        ]
+        remote_datasets = [
+            Dataset(namespace=self._get_namespace(scheme, remote_host, remote_port, path), name=path)
+            for path in remote_filepath
+        ]
+        print(remote_datasets)
```
Comment (phanikumv) [inline airflow/providers/sftp/operators/sftp.py]: please remove `print` here

**Review by phanikumv — COMMENTED** [review]
Can we add docs for this? how does a potential user know how to use the openlineage functionality along with sftp-operator?

File: airflow/providers/sftp/operators/sftp.py [hunk 2]
```diff
@@ -188,3 +191,85 @@ def execute(self, context: Any) -> str | list[str] | None:
...
+            local_filepath = [self.local_filepath]
+        else:
+            local_filepath = self.local_filepath
+        if isinstance(self.remote_filepath, str):
+            remote_filepath = [self.remote_filepath]
+        else:
+            remote_filepath = self.remote_filepath
+
+        local_datasets = [
+            Dataset(namespace=self._get_namespace(scheme, local_host, None, path), name=path)
```
Comment (utkarsharma2) [inline airflow/providers/sftp/operators/sftp.py:251]: I think this line will fail if the import failed `from openlineage.client.run import Dataset` right? Should we handle it better?

File: airflow/providers/sftp/operators/sftp.py [hunk 3]
```diff
@@ -188,3 +191,85 @@ def execute(self, context: Any) -> str | list[str] | None:
             raise AirflowException(f"Error while transferring {file_msg}, error: {str(e)}")
 
         return self.local_filepath
+
+    def get_openlineage_facets_on_start(self):
+        try:
```
Comment (utkarsharma2) [inline airflow/providers/sftp/operators/sftp.py]: just, curious is openlineage an optional package?

File: airflow/providers/sftp/operators/sftp.py [hunk 4]
```diff
@@ -188,3 +191,85 @@ def execute(self, context: Any) -> str | list[str] | None:
...
+                f"Failed to resolve local hostname. Using the hostname got by socket.gethostbyname() without resolution. {e}",  # noqa: E501
+                exc_info=True,
+            )
+
+        if self.sftp_hook is not None:
+            hook = self.sftp_hook
+        elif self.ssh_hook is not None:
+            hook = self.ssh_hook
+        else:
+            hook = SFTPHook(ssh_conn_id=self.ssh_conn_id)
```
Comment (utkarsharma2) [inline airflow/providers/sftp/operators/sftp.py]:
```suggestion
       hook = self.sftp_hook or self.ssh_hook or SFTPHook(ssh_conn_id=self.ssh_conn_id)
```

Comment (utkarsharma2) [reply airflow/providers/sftp/operators/sftp.py]: based on your preference, not mandatory :)

File: airflow/providers/sftp/operators/sftp.py [hunk 5]
```diff
@@ -188,3 +191,85 @@ def execute(self, context: Any) -> str | list[str] | None:
...
+
+        if self.remote_host is not None:
+            remote_host = self.remote_host
+        else:
+            remote_host = hook.get_connection(hook.ssh_conn_id).host
```
Comment (utkarsharma2) [inline airflow/providers/sftp/operators/sftp.py:220]: If the hook is initialized from `self.sftp_hook` will it still have `hook.ssh_conn_id`??

File: airflow/providers/sftp/operators/sftp.py [hunk 6]
```diff
@@ -188,3 +191,85 @@ def execute(self, context: Any) -> str | list[str] | None:
...
+        try:
+            remote_host = socket.gethostbyname(remote_host)
+        except Exception as e:
```
Comment (utkarsharma2) [inline airflow/providers/sftp/operators/sftp.py]: Is it possible to handle more specific exceptions?

File: tests/providers/sftp/operators/test_sftp.py [hunk 7]
```diff
@@ -478,3 +481,112 @@ def test_return_str_when_local_filepath_was_str(self, mock_get):
...
+            local_filepath="/path/local",
+            remote_filepath="/path/remote",
+            operation=operation,
+        )
+        lineage = task.get_openlineage_facets_on_start()
+
+        assert lineage.inputs == expected[0]
+        assert lineage.outputs == expected[1]
+
+    @pytest.mark.parametrize(
```
Comment (utkarsharma2) [inline tests/providers/sftp/operators/test_sftp.py:528]:
```
[
    (SFTPOperation.GET, (REMOTE_DATASET, LOCAL_DATASET)),
    (SFTPOperation.PUT, (LOCAL_DATASET, REMOTE_DATASET)),
],
```
I see this repeating maybe we can refactor it?

File: tests/providers/sftp/operators/test_sftp.py [hunk 8]
```diff
@@ -478,3 +481,112 @@ def test_return_str_when_local_filepath_was_str(self, mock_get):
         return_value = sftp_op.execute(None)
         assert isinstance(return_value, str)
         assert return_value == local_filepath
+
```
Comment (utkarsharma2) [inline tests/providers/sftp/operators/test_sftp.py:496]: Should we also add a test case for missing imports case? ex - `from openlineage.client.run import Dataset`

File: airflow/providers/sftp/operators/sftp.py (hunk 5 above)
Comment (mobuchowski) [reply airflow/providers/sftp/operators/sftp.py:220]: Yes: https://github.com/apache/airflow/blob/9276310a43d17a9e9e38c2cb83686a15656896b2/airflow/providers/sftp/hooks/sftp.py#L95

File: airflow/providers/sftp/operators/sftp.py (hunk 6 above)
Comment (mobuchowski) [reply airflow/providers/sftp/operators/sftp.py]: Yes. There are several possible socket exceptions here, but all are subclassing `OSError` with which I replaced `Exception`.

File: airflow/providers/sftp/operators/sftp.py (hunk 3 above)
Comment (mobuchowski) [reply airflow/providers/sftp/operators/sftp.py]: Yes - but this also means that if OpenLineage is not installed, this should never get called. So the try-except here is superfluous, and, as in other PRs I've moved to checking this in call site.

File: airflow/providers/sftp/operators/sftp.py (hunk 1 above)
Comment (mobuchowski) [reply airflow/providers/sftp/operators/sftp.py]: Done.

File: tests/providers/sftp/operators/test_sftp.py (hunk 8 above)
Comment (mobuchowski) [reply tests/providers/sftp/operators/test_sftp.py:496]: I think this should never happen - at least I do not see any reason or what would call those methods if OpenLineage is not installed.

**Comment by mobuchowski** [comment]
@phanikumv 

> Can we add docs for this? how does a potential user know how to use the openlineage functionality along with sftp-operator?

As discussed with @sunank200 in other PR, I've added description on how OL data returned from here looks like.

As for user-facing per-operator docs, I don't see reason to or what would be in them - the point of OL integration is that it works automatically without user action on particular operator level, as long as user has "global" OL configuration set up - which in most cases is just the transport.

File: tests/providers/sftp/operators/test_sftp.py (hunk 7 above)
Comment (mobuchowski) [reply tests/providers/sftp/operators/test_sftp.py:528]: Done.

**Review by phanikumv — APPROVED** [review]

**Pull request merged**